src/Data/Product/Algebra.agda

```diff
@@ -0,0 +1,278 @@
+
+  ×-zero : Zero _↔_ ⊥ _×_
+  ×-zero = ×-zeroˡ , ×-zeroʳ
+
+  -- × distributes over ⊎
+  ×-distribˡ-⊎ : _DistributesOverˡ_ {ℓ = ℓ} _↔_ _×_ _⊎_
+  ×-distribˡ-⊎ _ _ _ = inverse
+    (uncurry λ x → [ inj₁ ∘′ (x ,_) , inj₂ ∘′ (x ,_) ]′)
+    [ map₂ inj₁ , map₂ inj₂ ]′
+    Sum.[ cong′ , cong′ ]
+    (uncurry λ _ → Sum.[ cong′ , cong′ ])
+
+  ×-distribʳ-⊎ : _DistributesOverʳ_ {ℓ = ℓ} _↔_ _×_ _⊎_
+  ×-distribʳ-⊎ _ _ _ = inverse
+    (uncurry [ curry inj₁ , curry inj₂ ]′)
+    [ map₁ inj₁ , map₁ inj₂ ]′
+    Sum.[ cong′ , cong′ ]
+    (uncurry Sum.[ (λ _ → cong′) , (λ _ → cong′) ])
+
+  ×-distrib-⊎ : _DistributesOver_ {ℓ = ℓ} _↔_ _×_ _⊎_
+  ×-distrib-⊎ = ×-distribˡ-⊎ , ×-distribʳ-⊎
+
+------------------------------------------------------------------------
+-- Algebraic structures
+
+  ×-isMagma : IsMagma {ℓ = ℓ} _↔_ _×_
+  ×-isMagma = record
+    { isEquivalence = ↔-isEquivalence
+    ; ∙-cong        = ×-cong
+    }
+
+  ×-isSemigroup : IsSemigroup _↔_ _×_
+  ×-isSemigroup = record
+    { isMagma = ×-isMagma
+    ; assoc   = λ _ _ _ → Σ-assoc
+    }
+
+  ×-isMonoid : IsMonoid _↔_ _×_ ⊤
+  ×-isMonoid = record
+    { isSemigroup = ×-isSemigroup
+    ; identity    = ×-identityˡ , ×-identityʳ
+    }
+
+  ×-isCommutativeMonoid : IsCommutativeMonoid _↔_ _×_ ⊤
+  ×-isCommutativeMonoid = record
+    { isMonoid = ×-isMonoid
+    ; comm     = ×-comm
+    }
+
+  ⊎-×-isSemiringWithoutAnnihilatingZero : IsSemiringWithoutAnnihilatingZero _↔_ _⊎_ _×_ ⊥ ⊤
+  ⊎-×-isSemiringWithoutAnnihilatingZero = record
+    { +-isCommutativeMonoid = ⊎-isCommutativeMonoid ℓ
+    ; *-isMonoid            = ×-isMonoid
+    ; distrib               = ×-distrib-⊎
+    }
+
+  ⊎-×-isSemiring : IsSemiring _↔_ _⊎_ _×_ ⊥ ⊤
+  ⊎-×-isSemiring = record
+    { isSemiringWithoutAnnihilatingZero = ⊎-×-isSemiringWithoutAnnihilatingZero
+    ; zero                              = ×-zero
+    }
+
+  ⊎-×-isCommutativeSemiring : IsCommutativeSemiring _↔_ _⊎_ _×_ ⊥ ⊤
+  ⊎-×-isCommutativeSemiring = record
+    { isSemiring = ⊎-×-isSemiring
+    ; *-comm     = ×-comm
+    }
+------------------------------------------------------------------------
+-- Algebraic bundles
+
+  ×-magma : Magma (suc ℓ) ℓ
+  ×-magma = record
+    { isMagma = ×-isMagma
+    }
+
+  ×-semigroup : Semigroup (suc ℓ) ℓ
+  ×-semigroup = record
+    { isSemigroup = ×-isSemigroup
+    }
+
+  ×-monoid : Monoid (suc ℓ) ℓ
+  ×-monoid = record
+    { isMonoid = ×-isMonoid
+    }
+
+  ×-commutativeMonoid : CommutativeMonoid (suc ℓ) ℓ
+  ×-commutativeMonoid = record
+    { isCommutativeMonoid = ×-isCommutativeMonoid
+    }
+
+  ×-⊎-commutativeSemiring : CommutativeSemiring (suc ℓ) ℓ
+  ×-⊎-commutativeSemiring = record
+    { isCommutativeSemiring = ⊎-×-isCommutativeSemiring
+    }
+
+------------------------------------------------------------------------
+-- Some reordering lemmas
+
+ΠΠ↔ΠΠ : (P : A → B → Set p) → ((x : A) (y : B) → P x y) ↔ ((y : B) (x : A) → P x y)
+ΠΠ↔ΠΠ _ = inverse flip flip cong′ cong′
```

Comment:
Hmm the proofs below this point shouldn't live in `Data.Product.Algebra` but I'm at a bit of a loss about where to put them... Anyone have any suggestions?

Reply:
Well I've had a stab at moving them to "sensible" places. If no-one has any more feedback I'll merge it in next time I'm in the area.

Reply:
These places seem eminently sensible to me!  Note: the CHANGELOG hasn't been updated for that.